### ISSUE TITLE
Feature request: format R codes on save

### ISSUE THREAD

**Issue by yutannihilation** [rstudio/rstudio#2563]
Currently, RStudio can automatically formats codes on save a bit:

* Ensure the source files end with newline
* Strip trailing horizontal whitespace when saving

As [styler](https://github.com/r-lib/styler) reached 1.0, how about integrating the feature with RStudio? 
 Or, is it possible to let users to insert "before save hook"? 

In my understanding, "format on save" is a very common feature in modern editors/IDEs. For example, 
VS Code lets users to choose the formatter and when/whether to format the code or not:

https://code.visualstudio.com/docs/editor/codebasics#_formatting

For another example, Emacs doesn't have direct "format on save", but users can define `before-save-hook` to format codes:

https://github.com/dominikh/go-mode.el/blob/ff87a1e80945d958d9c348ce4d2e65a797b38a4b/go-mode.el#L1151

I also believe "format on save" is useful to soften the frustration from the incompleteness of "format on type" (e.g. #2371).

**Comment by ronblum** [comment]
@yutannihilation Thank you for the observations and suggestions. We'll consider them as we work on future versions of RStudio.

**Comment by kevinushey** [comment]
We've thought a little bit about allowing user-defined save hooks as well. The main difficulty is handling save hooks that misbehave -- e.g. we want to make sure they can easily be interrupted, and that we can report / disable save hooks that don't behave well.

I definitely agree this would be a powerful / useful feature; however, it's one that we'll have to think about carefully if we approach it.

**Comment by breckuh** [comment]
With Javascript/TypeScript and Python, I can't live without Prettier (https://github.com/prettier/prettier) and black (https://github.com/ambv/black). Formatting work is completely eliminated. Having an ultra reliable opinionated code formatted for R would be terrific.

**Comment by lorenzwalthert** [comment]
I think if you really want to use this feature you could use the RStudio Addin and enable saving after styling via the environment variable and assign the keyboard shortcut `Ctrl + S` to the Addin. See `?styler::style_active_file()` for details (and yes, we should probably have used an R option instead of an environment variable). It won't be a perfect substitution for the save hook (i.e. it won't play nicely with `devtools::load_all()` and friends) but it could abridge for now.

**Comment by lorenzwalthert** [comment]
On a related note, there is now a styler pre-commit hook: https://github.com/lorenzwalthert/pre-commit-hooks. Instead of *on save*, this is *on commit*, but maybe still worth considering.

**Comment by Kalaschnik** [comment]
Any news on this? Coming from VS Code where users can set `formatOnSave` and select their desired formater/linter over a variety of programming languages is a super convenient thing. It would be great if RStudio would allow users to set a similar option too; Maybe styler would be a good default here as well: https://github.com/r-lib/styler/issues/658

**Comment by Kalaschnik** [comment]
further activity bump!

**Comment by Kalaschnik** [comment]
I just discovered Hadley Wickham’s wonderful style guide (https://style.tidyverse.org/)! This is awesome work! Just as concise as ESLint and AirBnB style guides in the JS/Node world.  @hadley would love that feature!

**Comment by cigrainger** [comment]
I'm kind of lost as to why this is so difficult to pull off. Couldn't an addin do this? Right now I'm avoiding using RStudio for two major reasons: vim-mode has laggy escape and there's no format on save. The thing about format on save as opposed to precommit is that readability while working is important. In every other language I work in, and when I work in neovim or vscode, I hit save and everything snaps into place. It means I spend less cognitive load thinking about formatting and more time being productive. It's amazing how much code being formatted as you expect it to be helps with scanning and readability.

**Comment by brianthomaslewis** [comment]
Bumping for additional traffic. Adding my voice to the chorus that this would be a great feature to have included in RStudio.

**Comment by MattCowgill** [comment]
Adding my voice to the chorus here. 

Anecdotally, I see a lot of R users migrating to VS Code from the RStudio IDE. The absence of features like this accelerate the migration.

**Comment by njtierney** [comment]
Hi there just wanting to add that I would love to see this feature implemented - it would be a really nice addition and like others said, help reduce cognitive load because your code would be nicely styled consistently

**Comment by zWarMob** [comment]
*enters the chat*
Can I have 1 on-save formatter too please?

It has indeed been an indispensable feature in other languages and IDEs

**Comment by ssh352** [comment]
I'd like Rstudio to run "Reformat Code" on save. Rstudio formatter seems better than styler because it can break up long lines.

**Comment by joeDespres** [comment]
> Chiming in here too. This is a pretty necessary feature in 2024 (and well before that honestly).

💯

**Comment by kevinushey** [comment]
Thanks for the feedback everyone! We're now targeting this work for the following RStudio release, and so we plan to begin work in the coming months.

**Comment by lorenzwalthert** [comment]
Great news Kevin, if you consider an implementation with {styler}, please let me know if any questions arise. Having worked on [{styler}](https://styler.r-lib.org/) since 7 years and [{precommit}](https://lorenzwalthert.github.io/precommit/) as a spin-off project, I can think of the following relevant things:
- {styler} was quite slow in its initial release in 2017, but we managed to achieve around 5x speedup since then. Also, top level expression caching and file level caching makes repeated styling much faster. Format on save would benefit from it a lot I think. I have not properly examined how fast RStudios Format Code is compared to {styler}, but a quick test showed me that it's on the same order of magnitude.
- I also have not properly examined how {styler}’s results differ from Format Code but a quick test showed it produces notable differences. If people are using either one currently, activating Format on Save might interact with their usage patterns.
- I find very little documentation on Format Code style guide etc, whereas {styler} implements the tidyyverse style guide. I have interpreted that [quite strictly in the past](https://github.com/tidyverse/style/pull/174#issuecomment-909212890).
- {styler} still can’t break long lines, [the most popular feature request](https://github.com/r-lib/styler/issues/247) for a long time. 
- styler can be customised at various levels (create own style guide, use a third-party style guide, use non-default options for tidy verse style guide). A unified way to customise the RStudio Addin, the R API and third-party integrations (like pre-commit) does currently not exist but is an [open issue](https://github.com/r-lib/styler/issues/319).

Obviously up to you what to do with it, but I just thought I'd offer my two cents. I'd be excited if {styler} was integrated more tightly with RStudio, but I don't know if you consider that out of scope and prefer to just integrate the Format code RStudio command.

**Comment by ssh352** [comment]
It would be nice if styler and Rstudio Format code could finally be one tool like Ruff in Python or Rustfmt in Rust.  

R community is quite small and we don't need two separate formatters.

**Comment by davidbudzynski** [comment]
> It would be nice if styler and Rstudio Format code could finally be one tool like Ruff in Python or Rustfmt in Rust.
> 
> R community is quite small and we don't need two separate formatters.

There are many Python formatters so it's not like Ruff is the only one being used. Rustfmt is part of the language toolchain, so if you wanted something like that it would be the R devs you'd need to speak to, not RStudio or Styler devs

**Comment by ssh352** [comment]
> There are many Python formatters so it's not like Ruff is the only one being used. Rustfmt is part of the language toolchain, so if you wanted something like that it would be the R devs you'd need to speak to, not RStudio or Styler devs

That's true. Python has way more people working on formatters. But Ruff or black are what people really use nowadays (Ruff is much nicer). If R had only one formatter, I am just guessing: maybe that formatter could have be nicer and feature complete than both styler and Rstudio formatter now?

**Comment by lorenzwalthert** [comment]
RStudio's formatter was already there before we started developing {styler}, but we wanted a customisable, pure R implementation and R APIs (RStudio's formatter can only style the currently active file IIRC). {styler} has now many [third-party integrations](https://styler.r-lib.org/dev/articles/third-party-integrations.html) some enable styling in Jupyter Notebooks or VS Code or as a pre-commit hook. So there were different goals from the beginning, leading to different use cases. Also, as we can't go back in the past, the question you ask seems not particularly relevant to this issue. Of course, anyone who wants can build a much faster R formatter than {styler} if they build it in a faster language (or leverage existing formatting infrastructure written in another language), which could even be more feature complete. I'd be the first to use it 😀. But I don't see that happening because I don't think that's an important frontier in the R ecosystem at the moment or an easy thing to do.

**Comment by kevinushey** [comment]
Going to finally start taking a look at this -- sorry for taking so long. Here's a proposed implementation:

- Perform a "regular" save, so that the document on disk is in-sync with the editor buffer;
- Invoke a configured formatter on the file;
  - Should this be an external program? Someone could in theory use `R -s -e <command>` if they wanted to format using that program.
  - Or should this be an invocable R function registered in the session? We'd need to be more careful to handle errors and the like when that function is invoked, but this could be easier to use (e.g. testing / prototyping), and in theory that function could also itself just invoke some external program if needed.
- Re-sync the editor buffer contents with the formatted text contents.

Some questions:

- Should the formatter run on autosave as well (when enabled)? Or only when the user explicitly saves the file via a command / shortcut? I'm leaning towards the latter, just because having the document mutated without an explicit user gesture could feel surprising.

- How important are "format range" gestures for R? Does `styler` support formatting only some range of code on save?

**Comment by lorenzwalthert** [comment]
Great @kevinushey. To address your points, here's my take:
* I first thought `R -s -e <command>` seems a general solution, but one drawback I noted is that it won't play well with {renv} I think. In Python with poetry, I think it's common to define development dependencies to ensure everyone uses the same version of the formatter. Not exactly sure what the state of developer dependencies of {renv} is, but using an R function might be better suited for use with {renv}, if the R community goes towards declaring development dependencies with {renv} and use it also more for package development.
* R function could also invoke some external program, yes, but maybe `PATH` is not the same / must be configured in the R session and outside, which could cause friction.
* Maybe _format on autosave_ it could be an opt-in feature, but I in general think formatting should always lead to the same outcome, whether or not though auto-formatting or not.
* {styler}'s [RStudio Addin](https://styler.r-lib.org/reference/styler_addins.html) can format the selected code.

**Comment by t-kalinowski** [comment]
> Should this be an external program ... or should this be an invocable R function registered in the session?

My vote would be for an external program, so we leave the door open for the community to write a CLI app like `ruff` or `black` for R. Also, the R session can already be quite overburdened, I wouldn't clutter the main R process with the linter/formatter too. 

> Should the formatter run on autosave as well (when enabled)? Or only when the user explicitly saves the file

My vote would be to run only on explicit saves. It's disconcerting if the text changes while you are just looking at it.

> How important are "format range" gestures for R?

I think it's moderately important. This is an existing feature with `Cmd+Shift+A`, so hopefully that won't be removed. Two situation where this is useful:

- The user is contributing to a git repo where the formatter is not used by the original authors. The contributor wants to minimize the diff size, to format only the lines they changed, but not the full file. 
-  The user occasionally disagrees with the formatter, and rather than hack the formatter options, they prefer to apply the formatter in a very deliberate manner and then fix-up the code.

**Comment by kevinushey** [comment]
Thanks @lorenzwalthert -- do you have any strong opinions around how RStudio should invoke `styler` in this case? I'm assuming we'll want to use `styler::style_file()`?

- Is it possible to configure the default format style used by `styler` via a dotfile, or something similar? Or should we provide UI in RStudio to allow users to select the default style?

- Similarly, what scopes should RStudio select when formatting on save? Should we allow this to be configured via UI in RStudio, or would it be better to configure this externally (e.g. a `.styler` dotfile; R options; something else?)

**Comment by kevinushey** [comment]
> I think it's moderately important. This is an existing feature with Cmd+Shift+A, so hopefully that won't be removed. Two situation where this is useful:
> 
> - The user is contributing to a git repo where the formatter is not used by the original authors. The contributor wants to minimize the diff size, to format only the lines they changed, but not the full file.
> 
> - The user occasionally disagrees with the formatter, and rather than hack the formatter options, they prefer to apply the formatter in a very deliberate manner and then fix-up the code.

If I'm reading this correctly, this is more an argument for allowing customization of the formatter used for the "Reformat Code" command, as opposed to the behavior of a "format on save" feature? (And we should probably use the same formatter both on save, and for reformatting a snippet of code.)

I'll have to investigate this a bit more, I think.

**Comment by lorenzwalthert** [comment]
> I guess I was wondering -- if we allow for "format on save", should the whole file be formatted? Or just the lines that were edited? If the latter, how should we communicate which lines were edited to the reformatting tool? For example, imagine edits were made both at the start of a file, and at the end of a file -- how should we communicate those ranges?

The RStudio Addins supports formatting selected code, i.e. that functionality is [already there](https://github.com/r-lib/styler/blob/main/R/addins.R#L50) (kind of). One should note that formatting a range is more complicated (especially if start or end line are not the whole line, but just a part of it) and currently not very well unit-tested I think (not sure how addins unit testing works, or how we can capsule it), so before making this available to RStudio users, I guess someone would have to work on that (contributions appreciated).

**Comment by kevinushey** [comment]
The next daily builds will now have preliminary support for:

- Reformatting code / documents using the `styler` package, as well as
- Reformatting code on save using `styler`.

Configuration for the active formatter is available in the new "Formatting" tab:

<img width="678" alt="Screenshot 2024-08-02 at 1 54 25 PM" src="https://github.com/user-attachments/assets/36e81cb9-050d-46ec-bdcd-d8868868a0cc">


Reformatting on save is controlled in the "Saving" tab:

<img width="680" alt="Screenshot 2024-08-02 at 1 54 36 PM" src="https://github.com/user-attachments/assets/e8d4280a-6587-44a9-9d1e-4b2caa1ef0e2">


If you're interested in testing and providing feedback, these builds are available at https://dailies.rstudio.com. Please let us know what you think!

**Comment by kalaschnik** [comment]
Thanks @kevinushey for working on this! It is always the biggest bummer for me having no format-on-save in RStudio. This will also remove clutter from git diffs and will be a big gain for R's code consistency. Thank you ❤️

**Comment by lorenzwalthert** [comment]
Hi Kevin, thanks for implementing this, here are my observations and comments.

* toggle in Settings -> Code -> Saving -> 'Reformat Document on Save' does not seem to do anything. The setting also won't sync with the magic wand drop down where you can check or uncheck format on save. Is this the same setting or are these two formatters now? The magic wand (picture below) seems to work.

<img width="117" alt="Screenshot 2024-08-04 at 23 31 00" src="https://github.com/user-attachments/assets/8f62d00f-f60a-476f-aa7c-e357415369d9">

* Reformatting unsaved document does not seem to work.

My observations refer to RStudio-2024.10.0-daily-82 build for macOS.

**Comment by olivroy** [comment]
Similar to #14835, I think that Format on Save doesn't respect the existing line ending

**Comment by kevinushey** [comment]
> The magic wand (picture below) seems to work when I select format on save and auto-saving off and then save manually after some edits.

Thanks; I can fix this up.

> auto-saving on won't file the document. This might be desired, but then even manually triggering a saving won't format on save since the latest changes were already saved and there is nothing to save. Is that desired?

This should be fixed; any explicitly-requested saves should trigger formatting of the document.

> Reformatting unsaved document does not seem to work.

Can you clarify? Using "Reformat Selection" works for me, even in unsaved documents.

**Comment by kevinushey** [comment]
> I think that Format on Save doesn't respect the existing line ending.

I think this would need to be a feature of `styler`? Although I guess we could post-process the document after formatting.

**Comment by lorenzwalthert** [comment]
I am not sure what you mean with line ending, but we had an issue about support for different types of line endings: https://github.com/r-lib/styler/issues/396

**Comment by gtritchie** [comment]
Verified using build 2024.12.0-daily+375 on macOS.

**Comment by lorenzwalthert** [comment]
@kevinushey I just tried on the release version: After auto-safe with the following settings, hitting Cmd+S for saving, no reformatting is triggered. Maybe because _Save_ is not possible with no changes after last auto-saving?

<img width="327" alt="Image" src="https://github.com/user-attachments/assets/deda0899-132f-4b6f-aff9-f7c9c4b2817c" />
<img width="571" alt="Image" src="https://github.com/user-attachments/assets/1e081676-7a03-4135-ba29-37658b7b0dd2" />

**Comment by DavisVaughan** [comment]
> Save is not possible with no changes after last auto-saving

FWIW I'm fairly certain I've also seen this. If you haven't actually made any changes to the file, then if I remember right just hitting `Cmd + S` won't trigger the formatter to run. You have to "touch" the file first, and then resave.
